Mathlib/Algebra/Module/Submodule/IterateMapComap.lean

```diff
@@ -0,0 +1,104 @@
+`LinearMap.iterateMapComap f i m K = LinearMap.iterateMapComap f i (m + 1) K`,
+then for any `n`,
+`LinearMap.iterateMapComap f i n K = LinearMap.iterateMapComap f i (n + 1) K`.
+In particular, by taking `n = 0`, the kernel of `f` is contained in `K`
+(`LinearMap.ker_le_of_iterateMapComap_eq_succ`),
+which is a consequence of `LinearMap.ker_le_comap`.
+As a special case, if one can take `K` to be zero,
+then `f` is injective. This is the key result for establishing the strong rank condition
+for noetherian rings.
+
+The construction here is adapted from the proof in Djoković's paper
+*Epimorphisms of modules which must be isomorphisms* [djokovic1973].
+
+-/
+
+open Function Submodule
+
+namespace LinearMap
+
+variable {R M M₂ : Type*} [Semiring R] [AddCommMonoid M] [Module R M]
+  [AddCommMonoid M₂] [Module R M₂] (f i : M →ₗ[R] M₂)
+
+/-- The `LinearMap.iterateMapComap f i n K : Submodule R M` is
+`f⁻¹(i(⋯(f⁻¹(i(K)))))` (`n` times). -/
+def iterateMapComap (n : ℕ) := (fun K : Submodule R M ↦ (K.map i).comap f)^[n]
+
+/-- If `f(K) ≤ i(K)`, then `LinearMap.iterateMapComap` is not decreasing. -/
+theorem iterateMapComap_le_succ (K : Submodule R M) (h : K.map f ≤ K.map i) (n : ℕ) :
+    f.iterateMapComap i n K ≤ f.iterateMapComap i (n + 1) K := by
+  nth_rw 2 [iterateMapComap]
+  rw [iterate_succ', Function.comp_apply, ← iterateMapComap, ← map_le_iff_le_comap]
+  induction n with
+  | zero => exact h
+  | succ n ih =>
+    simp_rw [iterateMapComap, iterate_succ', Function.comp_apply]
+    calc
+      _ ≤ (f.iterateMapComap i n K).map i := map_comap_le _ _
+      _ ≤ (((f.iterateMapComap i n K).map f).comap f).map i := map_mono (le_comap_map _ _)
+      _ ≤ _ := map_mono (comap_mono ih)
+
+/-- If `LinearMap.iterateMapComap f i 0 K < LinearMap.iterateMapComap f i 1 K`,
+`f` is surjective, `i` is injective, then `LinearMap.iterateMapComap` is strictly increasing. -/
+theorem iterateMapComap_lt_succ (K : Submodule R M)
+    (h : f.iterateMapComap i 0 K < f.iterateMapComap i 1 K)
+    (hf : Surjective f) (hi : Injective i) (n : ℕ) :
+    f.iterateMapComap i n K < f.iterateMapComap i (n + 1) K := by
+  induction n with
+  | zero => exact h
+  | succ n ih =>
+    refine Ne.lt_of_le (fun H ↦ ?_)
+      (iterateMapComap_le_succ f i K (map_le_iff_le_comap.2 h.le) (n + 1))
+    rw [iterateMapComap, iterateMapComap, iterate_succ', iterate_succ'] at H
+    exact ih.ne (map_injective_of_injective hi (comap_injective_of_surjective hf H))
+
+/-- If `f(K) ≤ i(K)`, `f` is surjective, `i` is injective, and there exists some `m` such that
+`LinearMap.iterateMapComap f i m K = LinearMap.iterateMapComap f i (m + 1) K`,
+then for any `n`,
+`LinearMap.iterateMapComap f i n K = LinearMap.iterateMapComap f i (n + 1) K`.
+In particular, by taking `n = 0`, the kernel of `f` is contained in `K`
+(`LinearMap.ker_le_of_iterateMapComap_eq_succ`),
+which is a consequence of `LinearMap.ker_le_comap`. -/
+theorem iterateMapComap_eq_succ (K : Submodule R M) (h : K.map f ≤ K.map i)
```

Comment:
It is unlikely one will use this theorem, except for proving the next one, so I would try to prove only the second one.

Reply:
iterateMapComap cannot be strictly increasing, so the first two steps are equal, etc.

Reply:
> It is unlikely one will use this theorem, except for proving the next one, so I would try to prove only the second one.

I see. Will change it.

Reply:
Done. A superfluous condition is also removed.